README.md
Lambda Live Demo of Kana Tools for Java
=======================================
Code Powering the Live Demo of "Kana Tools for Java" on AWS Lambda and API Gateway

## Features
* Requires Java 8
* Live API is hosted on "serverless" architecture using Amazon AWS **Lambda** and **API Gateway**
* All information about HTTP requests made to API Gateway is passed via a **Lambda Proxy** JSON string `InputStream` to a function handler as which is executed on AWS Lambda

## Build
After cloning this repository, package is built using the Gradle wrapper

    cd lambda-kanatools-demo
    ./gradlew jar

After building the `jar`, it can then be uploaded to AWS Lambda as-is

# Endpoints
## `convertKana`
#### API Gateway Path
    GET /kana/convert

#### Lambda Handler
    com.mariten.lambda.ConvertKanaDemo::handleLambdaRequest

#### Query Params
| Name               | Type         | Description                                             | Required? |
| ------------------ |:------------:| ------------------------------------------------------- |:---------:|
| **`input_str`**    | String       | UTF-8 string to convert                                 | Yes       |
| **`conv_ops`**     | Positive Int | One or more flags defining which conversions to perform | Yes       |
| **`ignore_chars`** | String       | No characters in this string will be converted          | No        |

#### Result Format
```
{
    "success": true,
    "output_str": "<converted result string>"
}
```


HISTORY.md
v1.0.0 :: Jan 05 2017
======================
* Created AWS Lambda function handler that allows the execution of the `kanaConvert` function from [kanatools-java](https://github.com/mariten/kanatools-java) via serverless computing
    * Will be used to power a live demo of `kanaConvert`
* Wrote basic set of utility helper functions for reading in Lambda-Proxy formatted API Gateway requests, and fetching parameter values from its JSON structure
* Handle HTTP status codes in return value
    * `200` for success, `400` if invalid parameters were received, or `500` if unexpected exceptions occur


src/main/java/com/mariten/lambda/util/ApiGatewayHelpers.java
package com.mariten.lambda.util;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.io.BufferedReader;
import java.util.Arrays;
import java.util.List;

import com.amazonaws.services.lambda.runtime.LambdaLogger;
import org.json.simple.DeserializationException;
import org.json.simple.JsonArray;
import org.json.simple.JsonObject;
import org.json.simple.Jsoner;

public class ApiGatewayHelpers
{
    public static JsonObject parseRequest(InputStream api_request, LambdaLogger logger)
    {
        BufferedReader request_reader = null;
        try {
            request_reader = new BufferedReader(new InputStreamReader(api_request));
            // On Success, return parsed JSON structure
            return (JsonObject) Jsoner.deserialize(request_reader);

        } catch (IOException io_ex) {
            logger.log("[EXCEPTION] Error while reading stream: " + io_ex.getMessage());
        } catch (DeserializationException parse_ex) {
            logger.log("[EXCEPTION] Error parsing request JSON: " + parse_ex.getMessage());
        } catch (Exception ex) {
            logger.log("[EXCEPTION] Unexpected error: " + ex.getMessage());
        }

        // On Failure, return null after catching exception
        if (request_reader != null) {
            logger.log("    Failed request: " + request_reader.toString());
        }
        return null;
    }


    public static void respondSuccess(OutputStream api_response, JsonObject body_json)
    throws IOException
    {
        JsonObject success_json = new JsonObject();
        success_json.put("statusCode", "200");
        success_json.put("body", body_json.toJson());
        sendResponse(api_response, success_json);
    }


    public static void respondError(OutputStream api_response, int http_code, List<String> error_messages)
    throws IOException
    {
        JsonObject body_json = new JsonObject();
        body_json.put("success", false);
        body_json.put("errors", new JsonArray(error_messages));

        JsonObject error_json = new JsonObject();
        error_json.put("statusCode", Integer.toString(http_code));
        error_json.put("body", body_json.toJson());
        sendResponse(api_response, error_json);
    }
    public static void respondError(OutputStream api_response, int http_code, String error)
    throws IOException
    {
        respondError(api_response, http_code, Arrays.asList(error));
    }


    protected static void sendResponse(OutputStream api_response, JsonObject full_response)
    throws IOException
    {
        // Set HTTP headers for response
        JsonObject response_headers = new JsonObject();
        response_headers.put("content-type", "application/json");
        full_response.put("headers", response_headers);

        // Send response as an OutputStream to API Gateway
        OutputStreamWriter stream_writer = new OutputStreamWriter(api_response, "UTF-8");
        stream_writer.write(full_response.toJson());
        stream_writer.close();
    }


    public static String getStringQueryParam(JsonObject request_json, String param_name, String default_value)
    {
        JsonObject query_params = request_json.getMapOrDefault("queryStringParameters", null);
        if (query_params == null) {
            return default_value;
        }

        return query_params.getStringOrDefault(param_name, default_value);
    }


    public static int getIntegerQueryParam(JsonObject request_json, String param_name, int default_value)
    {
        JsonObject query_params = request_json.getMapOrDefault("queryStringParameters", null);
        if (query_params == null) {
            return default_value;
        }

        String raw_value = query_params.getStringOrDefault(param_name, null);
        if (raw_value == null) {
            return default_value;
        }

        try {
            return Integer.parseInt(raw_value);
        } catch (NumberFormatException num_ex) {
            return default_value;
        }
    }
}


src/main/java/com/mariten/lambda/ConvertKanaDemo.java
package com.mariten.lambda;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.LambdaLogger;
import com.mariten.kanatools.KanaConverter;
import org.json.simple.JsonObject;

import com.mariten.lambda.util.ApiGatewayHelpers;

public class ConvertKanaDemo
{
    public void handleLambdaRequest(InputStream     api_gateway_request,
                                    OutputStream    api_gateway_response,
                                    Context         lambda_context
    )
    throws IOException
    {
        LambdaLogger logger = lambda_context.getLogger();
        logger.log("Starting ConvertKanaDemo\n");

        // Parse request JSON structure received from API Gateway
        JsonObject request_json = ApiGatewayHelpers.parseRequest(api_gateway_request, logger);
        if (request_json == null) {
            ApiGatewayHelpers.respondError(api_gateway_response, 500, "Failure parsing request");
            return;
        }

        // Check for required params
        List<String> param_errors = new ArrayList<String>();
        //// Required: "input_str"
        String input_str = ApiGatewayHelpers.getStringQueryParam(request_json, "input_str", null);
        if (input_str == null) {
            param_errors.add("Query param [input_str] is required");
        }
        //// Required: "conv_ops"
        int conv_ops = ApiGatewayHelpers.getIntegerQueryParam(request_json, "conv_ops", 0);
        if (conv_ops < 1) {
            param_errors.add("Query param [conv_ops] is required, must be positive integer");
        }
        //// Optional: "ignore_chars"
        String ignore_chars = ApiGatewayHelpers.getStringQueryParam(request_json, "ignore_chars", "");

        // If any errors occurred, stop and return 400
        if (param_errors.size() > 0) {
            ApiGatewayHelpers.respondError(api_gateway_response, 400, param_errors);
            return;
        }

        // Convert string
        String output_str = KanaConverter.convertKana(input_str, conv_ops, ignore_chars);

        // Produce output JSON for response
        JsonObject response_json = new JsonObject();
        response_json.put("success", true);
        response_json.put("converted_str", output_str);

        // Send response and finish
        ApiGatewayHelpers.respondSuccess(api_gateway_response, response_json);
        logger.log("Finished ConvertKanaDemo\n");
    }
}
